Corona-Warn-App/src/main/java/de/rki/coronawarnapp/diagnosiskeys/download/KeyFileDownloader.kt

```diff
@@ -0,0 +1,303 @@
+            if (availCountry == null) {
+                Timber.w("Unknown location %s, assuming stale.", cachedHour.location)
+                return@filter true // It's stale
+            }
+
+            val availableDay = availCountry.hourData[cachedHour.day]
+            if (availableDay == null) {
+                Timber.d("Unknown day %s, assuming stale.", cachedHour.location)
+                return@filter true // It's stale
+            }
+
+            availableDay.none { time ->
+                cachedHour.hour == time
+            }
+        }
+        if (staleHours.isNotEmpty()) keyCache.delete(staleHours)
+
+        val nonStaleHours = cachedHours.minus(staleHours)
+        val missingHours = availableHours.mapNotNull {
+            it.toMissingHours(nonStaleHours)
+        }
+        Timber.tag(TAG).d("Downloading missing hours: %s", missingHours)
+
+        val hourDownloads = missingHours.flatMap { country ->
+            country.hourData.flatMap { (day, missingHours) ->
+                missingHours.map { missingHour ->
+                    async {
+                        val (keyInfo, path) = keyCache.createCacheEntry(
+                            location = country.country,
+                            dayIdentifier = day,
+                            hourIdentifier = missingHour,
+                            type = CachedKeyInfo.Type.COUNTRY_HOUR
+                        )
+
+                        return@async try {
+                            downloadKeyFile(keyInfo, path)
+                            keyInfo to path
+                        } catch (e: Exception) {
+                            Timber.tag(TAG).e(e, "Download failed: %s", keyInfo)
+                            null
+                        }
+                    }
+                }
+            }
+        }
+
+        Timber.tag(TAG).d("Waiting for %d missing hour downloads.", hourDownloads.size)
+        val downloadedHours = hourDownloads.awaitAll().filterNotNull()
+
+        downloadedHours.map { (keyInfo, path) ->
+            Timber.tag(TAG).d("Downloaded keyfile: %s to %s", keyInfo, path)
+            path
+        }
+
+        Unit
+    }
+
+    private suspend fun downloadKeyFile(keyInfo: CachedKeyInfo, saveTo: File) {
+        val validation = object : DiagnosisKeyServer.HeaderValidation {
+            override suspend fun validate(headers: Headers): Boolean {
+                var fileMD5 = headers.values("ETag").singleOrNull()
+//                var fileMD5 = headers.values("x-amz-meta-cwa-hash-md5").singleOrNull()
+//                if (fileMD5 == null) {
+//                    headers.values("x-amz-meta-cwa-hash").singleOrNull()
+//                }
+//                if (fileMD5 == null) { // Fallback
+//                    fileMD5 = headers.values("ETag").singleOrNull()
+//                }
+                fileMD5 = fileMD5?.removePrefix("\"")?.removeSuffix("\"")
+
+                return !legacyKeyCache.tryMigration(fileMD5, saveTo)
```

Comment:
Currently don't see a much better way to do the migration here, we can't totally hide it because we still need it to pass the target file. We also don't want the migration code to be hidden and not clutter our downloader, but not at all costs by shifting too many responsibilities too it. With the hook and the call here we have a quick way to "replace" the actual download. Abstracting this further brings little gain until we have more use cases for validation/hooks, imho. I've added a comment that explains the return value of `tryMigration` here.